qlib/Mapper.qm

```diff
@@ -930,13 +931,42 @@ Mapper mapv(DataMap);
         }
 
         #! maps a single field to the target
-        private nothing mapFieldIntern(reference h, string key, hash rec, *int do_list) {
+        /**
+          * Performs the actual mapping
+          *
+          * @param h the hash to be updated with the mapped key/value pair
+          * @param key the column name (hash key) to be mapped (target field)
+          * @param rec input record - either single record of hash of lists (bulk);
+          *                           to increase performance, the input type
+          *                           (single record vs. bulk) is determined by
+          *                           the @ref do_list parameter. In case of
+          *                           bulk input, all the lists are supposed to
+          *                           have the same length @ref list_size
+          * @param do_list - whether the input record @ref rec is single record or
+          *                  bulk format (hash of lists)
+          * @param list_size - size of the lists in case the input is in bulk
+          *                    format (all lists must have the same length,
+          *                    asserted inside the function).
+          * @param crec - optional hash to be added to each processed record
+          *               before actual mapping (useful basicly only with 'code'
+          *               mappings)
+          *
+          * @note it is assumed that list_size > 0 whenever do_list is Qore::True
+          */
+        private nothing mapFieldIntern(reference h,
+                                       string key,
+                                       hash rec,
+                                       bool do_list,
+                                       int list_size,
+                                       *hash crec) {
+            # FIXME: assert(!do_list || list_size > 0);
+            # FIXME can add the assert for equal length of the lists --PQ 22-Mar-2017
             hash m = mapc{key};
 
             # closure to get the current record hash from a hash of lists
             # rec is not bound here for performance reasons (so it will remain unlocked)
             code getrec = hash sub (hash rc, int offset) {
-                return map {$1.key: $1.value[offset] ?? $1.value}, rc.pairIterator();
+                return map {$1.key: $1.value.typeCode() == NT_LIST ? $1.value[offset] : $1.value}, rc.pairIterator();
```

Comment:
we need it for performance and compatibility reasons - this was the main reason it's there - removing it and adding `crec` as an additional argument has to have a justification because this change reduces performance and also breaks compatibility with the previous implementation